go-controller/pkg/ovn/egressip.go

```diff
@@ -0,0 +1,542 @@
+				namespace := obj.(*kapi.Namespace)
+				klog.V(5).Infof("EgressIP: %s stopped matching on namespace: %s", eIP.Name, namespace.Name)
+				if err := oc.deleteNamespaceEgressIP(eIP, namespace); err != nil {
+					klog.Errorf("error: unable to delete namespace handler for EgressIP: %s, err: %v", eIP.Name, err)
+				}
+			},
+		}, nil)
+	if err != nil {
+		return err
+	}
+	oc.egressIPNamespaceHandlerCache[getEgressIPKey(eIP)] = *h
+
+	if err := oc.updateEgressIPWithRetry(eIP); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (oc *Controller) deleteEgressIP(eIP *egressipv1.EgressIP) error {
+	if err := oc.releaseEgressIPs(eIP); err != nil {
+		return err
+	}
+
+	oc.egressIPNamespaceHandlerMutex.Lock()
+	defer oc.egressIPNamespaceHandlerMutex.Unlock()
+	delete(oc.egressIPNamespaceHandlerCache, getEgressIPKey(eIP))
+
+	oc.egressIPPodHandlerMutex.Lock()
+	defer oc.egressIPPodHandlerMutex.Unlock()
+	delete(oc.egressIPPodHandlerCache, getEgressIPKey(eIP))
+
+	namespaces, err := oc.kube.GetNamespaces(eIP.Spec.NamespaceSelector)
+	if err != nil {
+		return err
+	}
+	for _, namespace := range namespaces.Items {
+		if err := oc.deleteAllEgressPods(eIP, &namespace); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (oc *Controller) syncEgressIPs(eIPs []interface{}) {
+	oc.eIPAllocatorMutex.Lock()
+	defer oc.eIPAllocatorMutex.Unlock()
+	for _, eIP := range eIPs {
+		eIP, ok := eIP.(*egressipv1.EgressIP)
+		if !ok {
+			klog.Errorf("Spurious object in syncEgressIPs: %v", eIP)
+			continue
+		}
+		validAssignment := true
+		for _, eIPStatus := range eIP.Status {
+			validAssignment = false
+			eNode, exists := oc.eIPAllocator[eIPStatus.Node]
+			if !exists {
+				klog.Errorf("Allocator error: EgressIP: %s claims to have an allocation on a node which is unassignable for egress IP: %s", eIP.Name, eIPStatus.Node)
+				break
+			}
+			if !eNode.tainted {
+				ip := net.ParseIP(eIPStatus.EgressIP)
+				if ip == nil {
+					klog.Errorf("Allocator error: EgressIP allocation contains unparsable IP address: %s", eIPStatus.EgressIP)
+					break
+				}
+				if utilnet.IsIPv6(ip) && eNode.v6Subnet != nil {
+					if !eNode.v6Subnet.Contains(ip) {
+						klog.Errorf("Allocator error: EgressIP allocation: %s on subnet: %s which cannot host it", ip.String(), eNode.v6Subnet.String())
+						break
+					}
+				} else if !utilnet.IsIPv6(ip) && eNode.v4Subnet != nil {
+					if !eNode.v4Subnet.Contains(ip) {
+						klog.Errorf("Allocator error: EgressIP allocation: %s on subnet: %s which cannot host it", ip.String(), eNode.v4Subnet.String())
+						break
+					}
+				} else {
+					klog.Errorf("Allocator error: EgressIP allocation on node: %s which does not support its IP protocol version", eIPStatus.Node)
+					break
+				}
+				validAssignment = true
+			} else {
+				klog.Errorf("Allocator error: EgressIP: %s claims multiple egress IPs on same node: %s, will attempt rebalancing", eIP.Name, eIPStatus.Node)
+				break
+			}
+			eNode.tainted = true
```

Comment:
"tainted" is a bad name perhaps....

Essentially: an `EgressIP` can have multiple `EgressIP.Spec.EgressIPs` specified. We however don't want to allocate these egress IPs all on one node, we would - preferably - like to have a balanced and "HA" distribution of them. I am thus using this helper field `tainted` to enforce that condition, i.e: that the egress IPs a user defines for one `EgressIP` object, do not get assigned to the same node. 

I used "tainted" since this is naming convention used by the OpenShift scheduler to specify that "this node should be avoided for pod scheduling"...which is essentially what we are doing here, but with egress IPs.